### ISSUE TITLE
Can we use it for other datasets rather than davis 2016?

### ISSUE THREAD

**Issue by JawadTawhidi** [hkchengrex/vos-benchmark#3]
Hi, can we use it for other data sets rather than davis 2016? my dataset is ultrasound videos and is similar to davis 2016, may I use this script for calculating J and F values?

If not, would you please give me some clues on how can I calculate the J and F values for my results?

**Comment by JawadTawhidi** [comment]
I used it to calculate  J and F values on my ultrasound data set, during training process of my model when validation code is displaying the results after each epoch the highest J&F value was 66.737.

But when I used this repo for evaluation the J&F value is 66.4.


Now I am confused if this repo is working for evaluating my data set or not, would you please give me some clue?


One more thing I have to mention is that my data set has single object for most of the videos, but in some videos it has 3 to 4 objects as well.

**Comment by hkchengrex** [comment]
The script should work as long as the data is formatted as integer masks when read by PIL. Given that the difference in J&F is quite small, there might be slight implementation differences between your validation code and this repo. I do not know what those differences might be without knowing anything about your validation code.

**Comment by JawadTawhidi** [comment]
**As this result is very important for me, here I will attach my validation code to have your consideration.**


1. using this code I am preprocessing the data set:


class TestDAVIS(torch.utils.data.Dataset):
    def __init__(self, root, year, split):
        self.root = root
        self.year = year
        self.split = split
        self.init_data()

    
    
    #Original method
    def read_img(self, path):
        pic = Image.open(path).convert('L')

        transform = tv.transforms.ToTensor()
        return transform(pic)
    

    def read_mask(self, path):
        pic = Image.open(path).convert('P')


        transform = LabelToLongTensor()
        return transform(pic)

    def init_data(self):
        with open(os.path.join(self.root, 'ImageSets', self.year, self.split + '.txt'), 'r') as f:
            self.video_list = sorted(f.read().splitlines())
            print('--- DAVIS {} {} loaded for testing ---'.format(self.year, self.split))

    def get_snippet(self, video_name, frame_ids):
        img_path = os.path.join(self.root, 'JPEGImages', '480p', video_name)
        flow_path = os.path.join(self.root, 'JPEGFlows', '480p', video_name)
        mask_path = os.path.join(self.root, 'Annotations', '480p', video_name)
        imgs = torch.stack([self.read_img(os.path.join(img_path, '{:05d}.jpg'.format(i))) for i in frame_ids]).unsqueeze(0)
        flows = torch.stack([self.read_img(os.path.join(flow_path, '{:05d}.jpg'.format(i))) for i in frame_ids]).unsqueeze(0)
        given_masks = [self.read_mask(os.path.join(mask_path, '{:05d}.png'.format(0))).unsqueeze(0)] + [None] * (len(frame_ids) - 1)
        files = ['{:05d}.png'.format(i) for i in frame_ids]
        if self.split == 'test-dev':
            return {'imgs': imgs, 'flows': flows, 'given_masks': given_masks, 'files': files, 'val_frame_ids': None}
        masks = torch.stack([self.read_mask(os.path.join(mask_path, '{:05d}.png'.format(i))) for i in frame_ids]).unsqueeze(0)
        if self.year == '2016':
            masks = (masks != 0).long()
            given_masks[0] = (given_masks[0] != 0).long()
        return {'imgs': imgs, 'flows': flows, 'given_masks': given_masks, 'masks': masks, 'files': files, 'val_frame_ids': None}

    def get_video(self, video_name):
        frame_ids = sorted([int(file[:5]) for file in os.listdir(os.path.join(self.root, 'JPEGImages', '480p', video_name))])
        yield self.get_snippet(video_name, frame_ids)

    def get_videos(self):
        for video_name in self.video_list:
            yield video_name, self.get_video(video_name)


**2. This part is for changing the mask to tensor:**
import numpy as np
import torch


class LabelToLongTensor(object):
    def __call__(self, pic):
        if isinstance(pic, np.ndarray):
            label = torch.from_numpy(pic).long()
        elif pic.mode == '1':
            label = torch.from_numpy(np.array(pic, np.uint8, copy=False)).long().view(1, pic.size[1], pic.size[0])
        else:
            label = torch.ByteTensor(torch.ByteStorage.from_buffer(pic.tobytes()))
            if pic.mode == 'LA':
                label = label.view(pic.size[1], pic.size[0], 2)
                label = label.transpose(0, 1).transpose(0, 2).contiguous().long()[0]
                label = label.view(1, label.size(0), label.size(1))
            else:
                label = label.view(pic.size[1], pic.size[0], -1)
                label = label.transpose(0, 1).transpose(0, 2).contiguous().long()
        label[label == 255] = 0
        return label







**3. In this code I finalize the preprocessing and use official approach for evaluation:**


class Measure(object):
    #def __init__(self, model, val_set):
     #   self.model = model.cuda()
      #  self.val_set = val_set
    



    def __init__(self, model, val_set, cust_masks):
        self.model = model.cuda() if model is not None else None
        self.val_set = val_set
        self.cust_masks = cust_masks



    def cycle_dataset(self, mode):
       

        if mode == 'val':
            metrics_res = {}
            metrics_res['J'] = []
            metrics_res['F'] = []
            for video_name, video_parts in self.val_set.get_videos():
               
                for vos_data in video_parts:
                    imgs = vos_data['imgs'].cuda()
                    flows = vos_data['flows'].cuda()
                    masks = vos_data['masks'].cuda()
                    
                    

                    if self.model is not None:
                      
                      
                      # inference
                      vos_out = self.model(imgs, flows)
                      res_masks = vos_out['masks'][:, 1:-1].squeeze(2)


                    else:
                        # When the model is None
                      print('If this part of the code is running it means the masks from a folder is loading....')
                      for video_name, video_parts in self.cust_masks.get_videos():
                          for vos_data in video_parts:
                                # Load res_masks from custom_masks using CustMasks
                                 cust_masks = vos_data['masks'].cuda()
                                 
                                 res_masks = cust_masks[:, 1:-1].squeeze(2)
                                 res_masks = res_masks
                      

                    

                    gt_masks = masks[:, 1:-1].squeeze(2)
                    gt_masks= gt_masks


                    B, L, H, W = res_masks.shape
                    object_ids = numpy.unique(gt_masks.cpu()).tolist()
                    object_ids.remove(0)

                    # evaluate output
                    all_res_masks = numpy.zeros((len(object_ids), L, H, W))
                    all_gt_masks = numpy.zeros((len(object_ids), L, H, W))
                    for k in object_ids:
                        res_masks_k = copy.deepcopy(res_masks).cpu().numpy()
                        res_masks_k[res_masks_k != k] = 0
                        res_masks_k[res_masks_k != 0] = 1
                        all_res_masks[k - 1] = res_masks_k[0]
                        gt_masks_k = copy.deepcopy(gt_masks).cpu().numpy()
                        gt_masks_k[gt_masks_k != k] = 0
                        gt_masks_k[gt_masks_k != 0] = 1
                        all_gt_masks[k - 1] = gt_masks_k[0]

                    # calculate scores
                    j_metrics_res = numpy.zeros(all_gt_masks.shape[:2])
                    f_metrics_res = numpy.zeros(all_gt_masks.shape[:2])
                    for i in range(all_gt_masks.shape[0]):
                        j_metrics_res[i] = metrics.db_eval_iou(all_gt_masks[i], all_res_masks[i])
                        f_metrics_res[i] = metrics.db_eval_boundary(all_gt_masks[i], all_res_masks[i])
                        [JM, _, _] = metrics.db_statistics(j_metrics_res[i])
                        metrics_res['J'].append(JM)
                        [FM, _, _] = metrics.db_statistics(f_metrics_res[i])
                        metrics_res['F'].append(FM)

                        print('Video: {}, J: {:.5f}, F: {:.5f}'.format(video_name, JM, FM))


            # gather scores
            J, F = metrics_res['J'], metrics_res['F']
            final_mean = (numpy.mean(J) + numpy.mean(F)) / 2.
           # print('J&F score: {:.5f}'.format(final_mean))

            print('J&F score: {:.5f} (J: {:.5f}, F: {:.5f})'.format(final_mean, numpy.mean(J), numpy.mean(F)))
            return final_mean

**Comment by JawadTawhidi** [comment]
One more thing I have to mention again is that my data set contains single objects for most of the videos but for some videos it has multiple objects as well. That is why I used the davis 2017 approach also(The one you suggested in the deva repo, using that script the result was 66.5, while using this script it was 66.4 and using my validation it is 66.7)

**Comment by hkchengrex** [comment]
Hello. I cannot really look at all these code. I can help if there are specific questions regarding this repo.

**Comment by JawadTawhidi** [comment]
Ok, thank you.

But it seems this script is not working when the ground masks are normalized(Every pixel divided by 255). It only works when both ground masks and predicted masks are in same format(Without color and unnormalized)

**Comment by JawadTawhidi** [comment]
I used this code to test the results generated for my dataset using another model, again in that other model the J&F value which is provided by validation code after training is 69.027, but when I tested the results using this script it is 64.2. 

Why this much differences?
